grakn-engine/src/main/java/ai/grakn/engine/lock/ZookeeperLock.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Grakn - A Distributed Semantic Database
+ * Copyright (C) 2016  Grakn Labs Limited
+ *
+ * Grakn is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Grakn is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Grakn. If not, see <http://www.gnu.org/licenses/gpl.txt>.
+ */
+
+package ai.grakn.engine.lock;
+
+import ai.grakn.engine.tasks.manager.ZookeeperConnection;
+import ai.grakn.exception.EngineStorageException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import org.apache.curator.framework.recipes.locks.InterProcessSemaphoreMutex;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+/**
+ * <p>
+ *     A fully distributed lock that is globally synchronous. No two engines or tasks will think they hold the same lock.
+ * </p>
+ *
+ * @author alexandraorth
+ */
+public class ZookeeperLock implements Lock {
```

Comment:
I don't think I'm planning on implementing the other lock functionalities because we wont be using those. We could, of course, but way in the future when we have time.